### PR TITLE
Expose checkpointing in C API

### PR DESCRIPTION
**Pull request by szkarpinski**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
This PR exposes the checkpointing functionality in C API. This is a preparation for checkpointing support in TF plugin. The following changes were made:

1. `enable_checkpointing` flag is now included in serialized pipeline and is set accordingly on deserialization.
2. `daliGetSerializedCheckpoint` and `daliRestoreFromSerializedCheckpoint` calls, equivalent to `Pipeline::RestoreFromSerializedCheckpoint` and `Pipeline::daliGetSerializedCheckpoint` were added.

### Affected modules and functionalities:

- `dali.proto` and pipeline (de)serialization was changed to preserve `enable_checkpointing`
- C API was extended with two new functions

### Key points relevant for the review:
Are the signatures of the introduced functions aligned with our preferences and conventions?

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3783

### CODE REVIEW

File: dali/c_api/c_api.cc [hunk 1]
```diff
@@ -803,3 +804,28 @@ int daliPreallocatePinnedMemory(size_t bytes) {
     return -1;
   }
 }
+
+void daliGetSerializedCheckpoint(daliPipelineHandle_t pipe_handle,
+                                 daliExternalContextCheckpoint *external_context,
```
Comment (stiepan) [inline dali/c_api/c_api.cc:809]: The external_context could be const, right? I'd add a check for it being nullptr too.

**Review by stiepan — APPROVED** [review]

**Review by klecki — APPROVED** [review]

**Comment by szkarpinski** [comment]
@mzient You're right, at some point I was using RunCpu+RunGpu and forgot that I've changed it. I think we don't need `daliPrefetchUniform` here because prefetch_queue_depth is set to 1. Is that right?

File: dali/c_api/c_api.cc (hunk 1 above)
Comment (szkarpinski) [reply dali/c_api/c_api.cc]: Fixed, thank you!

**Comment by szkarpinski** [comment]
!build

**Pull request merged**